lib/sequence.js

```diff
@@ -1,86 +1,59 @@
 'use strict';
 const isPromise = require('is-promise');
-const autoBind = require('auto-bind');
-const AvaError = require('./ava-error');
 
 class Sequence {
-	constructor(tests, bail) {
-		if (!tests) {
-			throw new Error('Sequence items can\'t be undefined');
+	constructor(runnables, bail) {
+		if (!Array.isArray(runnables)) {
+			throw new TypeError('Expected an array of runnables');
 		}
 
-		this.results = [];
-		this.passed = true;
-		this.reason = null;
-		this.tests = tests;
+		this.runnables = runnables;
 		this.bail = bail || false;
-
-		autoBind(this);
 	}
+
 	run() {
-		const length = this.tests.length;
+		const iterator = this.runnables[Symbol.iterator]();
 
-		for (let i = 0; i < length; i++) {
-			// If last item failed and we should bail, return results and stop
-			if (this.bail && !this.passed) {
-				return this._results();
-			}
+		let allPassed = true;
+		const runNext = () => {
+			let promise;
 
-			const result = this.tests[i].run();
+			for (let next = iterator.next(); !next.done; next = iterator.next()) {
```

Comment:
Because of promises.  The alternative is maintaining an `index` variable to resume the iteration. That's what iterators do all by themselves though.